### DISCUSSION TITLE
p51 Minimal Version Selectionの実例について

### DISCUSSION THREAD

**Discussion by hajipy** [budougumi0617/go_todo_app#90] [Ideas]
- マイパッケージ
  - パッケージA
    - パッケージX v1.0.1
    - パッケージY v1.1.0
  - パッケージB
    - パッケージX v1.0.8
    - パッケージY v1.2.3

この場合にパッケージXのバージョンは「v1.0.1」が選ばれると記載されておりますが、「v1.0.8」が選ばれるのが正しい挙動ではないかと思います。

参考URLとして記載されていた https://research.swtch.com/vgo-mvs を確認したところ、 Algorithm 1: Construct Build Listのrough listにてパッケージDの1.3と1.4が残っていますが、final listでは1.4のみとなっています。

また、オライリーの「初めてGo言語」の「9.4.3 ミニマムバージョン選択」にも「v1.1.0とv1.2.0とv1.2.3に依存していた場合、v1.2.3が選択される」という旨の記載がありました。

Minimalというのは「複数の候補があった場合、その中で最小のものを選ぶ」ということではなく、「必要最低限のバージョンよりも新しいバージョンが公開されていても、必要最低限のバージョンを選ぶ」ということかと思います。
例えば、パッケージXのv1.1.0が公開されていた場合、セマンティックバージョニングのルールを守っているのであれば、これはv1.0.1やv1.0.8と互換性がありつつも、新しい機能を追加したバージョンとなります。
JavaScriptのNPMなどはMaximal Version Selectionを採用しており、「互換性があるのだから新しい方がいいだろう」と判断して、このv1.1.0を選びます。
Goで採用しているMinimal Version Selectionは、より保守的で「その新機能はどのパッケージも必要としていないので古いバージョンでいいだろう」と判断してv1.0.8を選ぶ、ということだと理解しています。

**Comment by budougumi0617** [comment]
ありがとうございます！！！たしかにご指摘のとおりでした。解説までありがとうございます 🙇 

エラッタに追記させていただきました
https://github.com/budougumi0617/go_todo_app/releases/tag/v1.1.1
